File: Scala/IdBench.scala
// JAVA_OPTS="-Xmx4G" scala IdBench.scala
// or
// scalac IdBench.scala; JAVA_OPTS="-Xmx4G" scala

object IdBench {
  def main(args: Array[String]) {
    val n = 1<<25
    val m = 8
    val niters = n*m
    var sum:Double = 0.0
    var idx:Int = 0

    // initialize data
    val data = Array.tabulate(n) {i => ((i+1)%n).toDouble}

    // main loop
    val st = System.nanoTime()
    for (i <- 0 until niters) {
      idx = data(data(data(data(idx).toInt).toInt).toInt).toInt
      sum += idx
    }
    val et = System.nanoTime() - st

    // stats
    val a = et/niters.toDouble
    println(f"Elapsed [msec]: " + (et*1e-6))
    println(f"Per iteration [nsec]: $a")
    println(f"sum=$sum")
    val ref = m.toDouble * n.toDouble * (n.toDouble-1.0) / 2.0.toDouble;
    if ( (sum-ref).abs < 1e-6) {
      println(f"Failed to validate: sum should be $ref")
    } else {
      println("Validation passed!")
    }
  }
}

File: C/idbench.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static double getuptime_nsec(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    return (double)ts.tv_sec*1e9 + (double)ts.tv_nsec;
}

int main()
{
  int n = 1<<25;
  int m = 8;
  int niters = n * m;
  double *data;
  double st, et;
  double sum=0.0;
  data = malloc(n*sizeof(double));

  // intialize data
  for (int i=0; i<n; i++) data[i] = (double)((i+1)%n);

  // main loop
  sum = 0.0;
  int idx = 0;
  st = getuptime_nsec();
  for (int i=0; i< niters; i++) {
    idx = data[(int)data[(int)data[idx]]];
    sum += idx;
  }
  et = getuptime_nsec() - st;

  // stats
  printf("Elapsed[msec]: %lf\n", et*1e-6);
  printf("Per iteration [nsec]: %lf\n", et/(double)niters);
  printf("sum=%lf\n", sum);

  double ref = (double)m*(double)n*(double)(n-1)/2.0;
  if((sum-ref) != 0.0) // rounding?
    printf("Failed to validate: sum should be %lf\n", ref);
  else
    printf("Validation passed!\n");

  return 0;
}
